test/updateFiles.test.ts

```diff
@@ -0,0 +1,244 @@
+  })
+
+  it('Can validate paths', async () => {
+    const invalidPaths = [
+      '/',
+      '/ bad/space',
+      'has/no/root/slash',
+      '/too/many//slashes'
+    ]
+
+    for (const path of invalidPaths) {
+      await agent
+        .post('/api/v3/updateFiles')
+        .send({
+          repoId,
+          timestamp: 0,
+          paths: {
+            [path]: { text: 'content' }
+          }
+        })
+        .expect(isErrorResponse(400, `Invalid path '${path}'`))
+    }
+  })
+
+  it('Can validate file data', async () => {
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: 0,
+        paths: {
+          '/file': { wrong: 'shape' }
+        }
+      })
+      .expect(isErrorResponse(400, `Expected null at .paths["/file"]`))
+  })
+
+  it('Can write file', async () => {
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          '/file': { text: 'content' }
+        }
+      })
+      .expect(isUpdateFilesResponse)
+  })
+
+  it('Can update file', async () => {
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          '/file': { text: 'content' }
+        }
+      })
+      .expect(isUpdateFilesResponse)
+  })
+
+  it('Can write file with directory', async () => {
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          '/dir/file': { text: 'content' }
+        }
+      })
+      .expect(isUpdateFilesResponse)
+  })
+
+  it('Cannot write file where there is a directory', async () => {
+    const filePath = '/dir'
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          [filePath]: { text: 'content' }
+        }
+      })
+      .expect(
+        isErrorResponse(
+          422,
+          `Unable to write file '${filePath}'. ` +
+            `Existing document is not a file.`
+        )
+      )
+  })
+
+  it('Cannot write file where the directory is a file', async () => {
+    const filePath = '/file/file'
+    const directoryPath = '/file'
+
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          [filePath]: { text: 'content' }
+        }
+      })
+      .expect(
+        isErrorResponse(
+          422,
+          `Unable to write files under '${directoryPath}'. ` +
+            `Existing document is not a directory.`
+        )
+      )
+  })
+
+  it('Can delete file', async () => {
+    await agent
+      .post('/api/v3/updateFiles')
+      .send({
+        repoId,
+        timestamp: repoTimestamp,
+        paths: {
+          '/file': null
+        }
+      })
+      .expect(isUpdateFilesResponse)
```

Comment:
This should also create the file before deleting it. (No interdependencies between tests)